src/qibo/models/iqae.py

```diff
@@ -0,0 +1,385 @@
+            T: maximum number of rounds to achieve the desired absolute error
+        Returns:
+            L_max, L_min: the maximum and minimum possible error which could be returned
+            on a given iteration
+        """
+        x = np.linspace(0, np.pi, 10000)
+        x_domain = [(x >= 0) & (x <= 1.0 + 1 / 10 / N_shots)]
+        alpha = self.alpha
+
+        def h_calc_CP(N_total_plus, N_total_shots, alpha, T):
+            """Calculates the h function"""
+            a_min, a_max = self.clopper_pearson(
+                N_total_plus, N_total_shots, alpha=(alpha / T)
+            )
+            return np.abs(np.arccos(1 - 2 * a_max) - np.arccos(1 - 2 * a_min)) / 2
+
+        y = [
+            (lambda t: h_calc_CP(int(t * N_shots), N_shots, alpha, T))(t)
+            for t in x[tuple(x_domain)]
+        ]
+        L_max = np.max(y) / (2 * np.pi)
+        L_min = np.min(y) / (2 * np.pi)
+        return L_max, L_min
+
+    def calc_L_range_CH(self, N_shots, T):
+        """
+        Calculate the confidence interval for the Chernoff method
+        Args:
+            N_shots: number of shots
+            T: maximum number of rounds to achieve the desired absolute error
+        Returns:
+            L_max, L_min: the maximum and minimum possible error which could be returned
+            on a given iteration
+        """
+        L_max = (
+            np.arcsin((2 / (N_shots) * np.log(2 * T / self.alpha)) ** (1 / 4))
+            / 2
+            / np.pi
+        )
+        L_min = np.arcsin(np.sin(L_max) ** 2)
+        return L_max, L_min
+
+    def find_next_k(self, K_i, up_i, theta_l, theta_u, r=2):
+        """Find the largest integer K such that the interval K*[theta_l,theta_u]
+        lies completely in [0, pi] or [pi, 2pi].
+
+        Args:
+            k: the current power of the Q operator
+            up_i: boolean flag of whether theta_interval lies in the
+                upper half-circle [0, pi] or in the lower one [pi, 2pi]
+            theta_l: the current lower limit of the confidence interval for the angle theta
+            theta_u: the current upper limit of the confidence interval for the angle theta
+
+        Returns:
+            The next power k, and boolean flag for the extrapolated interval
+        """
+        K_max = int(1 / (2 * (theta_u - theta_l)))
+        K = K_max - (K_max - 2) % 4
+        while K >= r * K_i:
+            theta_min = K * theta_l - int(K * theta_l)
+            theta_max = K * theta_u - int(K * theta_u)
+            if (
+                theta_max <= 1 / 2
+                and theta_min <= 1 / 2
+                and int(K * theta_u) == int(K * theta_l)
+            ):
+                up = True
+                return (K, up)
+            elif (
+                theta_max >= 1 / 2
+                and theta_min >= 1 / 2
+                and int(K * theta_u) == int(K * theta_l)
+            ):
+                up = False
+                return (K, up)
```

Comment:
Thanks for the suggestion but I think there was no problem for the specific case you mentioned because the function returns `up_i` instead of `up` in that case. Anyhow, thanks to that suggestion I found a more compact way to do the same. The corrected function will read as:

    def find_next_k(self, K_i, up_i, theta_l, theta_u, r=2):
        """Find the largest integer K such that the interval K*[theta_l,theta_u]
        lies completely in [0, pi] or [pi, 2pi].

        Args:
            K_i: the current K such K=4k+2, where k is the power of the Q operator
            up_i: boolean flag of whether theta_interval lies in the
                upper half-circle [0, pi] or in the lower one [pi, 2pi]
            theta_l: the current lower limit of the confidence interval for the angle theta
            theta_u: the current upper limit of the confidence interval for the angle theta
            r: lower bound for K

        Returns:
            The next power K_i, and boolean flag for the extrapolated interval
        """
        K_max = int(1 / (2 * (theta_u - theta_l)))
        K = K_max - (K_max - 2) % 4
        while K >= r * K_i:
            theta_min = K * theta_l - int(K * theta_l)
            theta_max = K * theta_u - int(K * theta_u)
            if int(K * theta_u) == int(K * theta_l):
                if(theta_max <= 1 / 2 and theta_min <= 1 / 2):
                    up=True
                    return (K,  up)
                elif(theta_max >= 1 / 2 and theta_min >= 1 / 2):
                    up=False
                    return (K,  up)
            K -= 4
        return (K_i, up_i)